sui_core/src/consensus_adapter.rs

```diff
@@ -166,19 +116,113 @@ impl ConsensusSubmitter {
             .await
             .expect("Failed to notify consensus listener");
 
-        // TODO [issue #1452]: We are re-creating a connection every time. This is wasteful but does not
-        // require to take self as a mutable reference.
-        Self::reconnect(self.consensus_address, self.buffer_size)
-            .await?
-            .sink()
-            .send(bytes)
-            .await
-            .map_err(|e| SuiError::ConsensusConnectionBroken(e.to_string()))?;
+        // Check if this authority submits the transaction to consensus.
+        if Self::should_submit(certificate) {
+            // TODO [issue #1452]: We are re-creating a connection every time. This is wasteful but
+            // does not require to take self as a mutable reference.
+            Self::reconnect(self.consensus_address, self.buffer_size)
+                .await?
+                .sink()
+                .send(bytes)
+                .await
+                .map_err(|e| SuiError::ConsensusConnectionBroken(e.to_string()))?;
+        }
 
         // Wait for the consensus to sequence the certificate and assign locks to shared objects.
         timeout(self.max_delay, receiver)
             .await
             .map_err(|e| SuiError::ConsensusConnectionBroken(e.to_string()))?
-            .expect("Chanel with consensus listener dropped")
+            .expect("Channel with consensus listener dropped")
+    }
+}
+
+/// This module interfaces the consensus with Sui. It receives certificates input to consensus and
+/// notify the called when they are sequenced.
+pub struct ConsensusListener {
+    /// The authority state to execute shared-object transactions.
+    state: Arc<AuthorityState>,
+    /// Receive messages input to the consensus.
+    rx_consensus_input: Receiver<ConsensusInput>,
+    /// Receive consensus outputs.
+    rx_consensus_output: Receiver<ConsensusOutput>,
+    /// Keep a map of all consensus inputs that are currently being sequenced.
+    pending: HashMap<ConsensusTransactionDigest, Vec<Replier>>,
+}
+
+impl ConsensusListener {
+    /// Spawn a new consensus adapter in a dedicated tokio task.
+    pub fn spawn(
+        state: Arc<AuthorityState>,
+        rx_consensus_input: Receiver<ConsensusInput>,
+        rx_consensus_output: Receiver<ConsensusOutput>,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            Self {
+                state,
+                rx_consensus_input,
+                rx_consensus_output,
+                pending: HashMap::new(),
```

Comment:
Advice: has a max size for this table -- it can be large, and after that return an error if there are too many pending things.

Reply:
https://github.com/MystenLabs/sui/issues/1680

Reply:
Will do in a separate PR